### DISCUSSION TITLE
Require a different PHP version for some classes of a library

### DISCUSSION THREAD

**Discussion by GromNaN** [composer/composer#11667] [Q&A / Support]
I maintain a library that needs to keep support for PHP 7.4 (for the time being).
For a new feature, I need to use the new PHP 8.1 features (union and intersection type, and named arguments).

I would like my users to have a relevant warning message when they try to use this feature with an older version of PHP.

I've thought of adding the following code to the header of all my classes, but it doesn't work because the PHP compilation makes an error before I can display the message.

```php
<?php

if (PHP_VERSION_ID < 80100) {
    throw new \LogicException('The feature requires PHP 8.1');
}

class FeatureThatRequiresPHP81 {}
```

I don't think there's a mechanism in composer for this, but maybe a way to hook the class loader to emit the error before loading the file.

**Comment by fredden** [comment]
You should list the supported versions of PHP in the `require` section of `composer.json`. It sounds like you're dropping support for PHP versions less than 8.1 by employing these language features, so you should change the `require.php` restraint accordingly (probably "`~8.1.0 || ~8.2.0`"). Composer will then help you by only offering this new version to those who have compatible versions of PHP. If you're using semantic versioning, this sounds like a new major version. You can still release bug-fixes for previous versions by incrementing the appropriate version number part.

**Reply by GromNaN** [reply]
The new versions of the library must still accept PHP 7.4, so I cannot change the required php version.

But correct, I can maintain 2 branches in parallel. One with PHP 7.4 support and the other with the new feature requiring PHP 8.1.
